File: android/app/src/main/java/com/algonit/algo/features/chat/data/model/StructuredData.kt
package com.algonit.algo.features.chat.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject

/**
 * Structured data attached to assistant responses for rich display.
 * Used to render metrics cards, charts, and action confirmation details.
 *
 * @property type The data type: "performance_summary", "action_confirmation", "action_result", etc.
 * @property metrics A list of metric data points for charts and metric cards.
 * @property chartType The chart visualization type: "bar", "line", "pie", etc.
 * @property timeRange The time period the data covers: "today", "week", "month".
 * @property action The action identifier for confirmation/result types (e.g., "PAUSE_CAMPAIGN").
 * @property status The status of an action result: "success", "failed", "pending".
 * @property target Additional target-specific details for the action, such as campaign info.
 */
@Serializable
data class StructuredData(
    val type: String,
    val metrics: List<JsonObject>? = null,
    @SerialName("chart_type") val chartType: String? = null,
    @SerialName("time_range") val timeRange: String? = null,
    val action: String? = null,
    val status: String? = null,
    val target: JsonObject? = null
) {
    /**
     * Whether this structured data represents a performance metrics summary.
     */
    val isPerformanceSummary: Boolean
        get() = type == "performance_summary"

    /**
     * Whether this structured data represents an action requiring user confirmation.
     */
    val isActionConfirmation: Boolean
        get() = type == "action_confirmation"

    /**
     * Whether this structured data represents the result of an executed action.
     */
    val isActionResult: Boolean
        get() = type == "action_result"

    /**
     * Whether this structured data has chart-renderable metrics.
     */
    val hasChart: Boolean
        get() = !metrics.isNullOrEmpty() && chartType != null
}

File: ios/Algo/Features/Chat/Models/StructuredData.swift
// StructuredData.swift
// Algo
//
// Structured data model for rich content returned by the assistant,
// such as performance summaries, charts, and action confirmation details.

import Foundation

// MARK: - StructuredData

struct StructuredData: Codable, Sendable, Equatable {

    /// The type of structured content (e.g. "performance_summary", "action_confirmation", "action_result").
    let type: String

    /// Array of metric dictionaries for data visualization (charts, tables).
    /// Each entry is a flexible key-value map to accommodate varying metric shapes.
    let metrics: [[String: JSONValue]]?

    /// Suggested chart visualization type (e.g. "bar", "line", "pie").
    let chartType: String?

    /// Time range label for the data (e.g. "today", "last_7_days").
    let timeRange: String?

    /// The action being confirmed or executed (e.g. "PAUSE_CAMPAIGN").
    let action: String?

    /// The status of a completed action (e.g. "success", "failed").
    let status: String?

    /// Target entity details for action confirmations.
    let target: [String: JSONValue]?

    /// Campaign ID for action results referencing a specific campaign.
    let campaignId: String?

    enum CodingKeys: String, CodingKey {
        case type
        case metrics
        case chartType = "chart_type"
        case timeRange = "time_range"
        case action
        case status
        case target
        case campaignId = "campaign_id"
    }
}

// MARK: - StructuredData Type Constants

extension StructuredData {

    /// Known structured data types returned by the backend.
    enum DataType {
        static let performanceSummary = "performance_summary"
        static let actionConfirmation = "action_confirmation"
        static let actionResult = "action_result"
        static let leadList = "lead_list"
        static let campaignList = "campaign_list"
        static let analyticsReport = "analytics_report"
    }

    /// Whether this structured data represents a chart-renderable dataset.
    var isChartData: Bool {
        chartType != nil && metrics != nil
    }

    /// Whether this structured data represents an action requiring user confirmation.
    var isActionConfirmation: Bool {
        type == DataType.actionConfirmation
    }

    /// Whether this structured data represents the result of a completed action.
    var isActionResult: Bool {
        type == DataType.actionResult
    }

    /// The chart type as a typed enum, if recognized.
    var chartTypeEnum: ChartType? {
        guard let chartType else { return nil }
        return ChartType(rawValue: chartType)
    }

    /// Extracts a flat list of metric display entries from the raw JSON metrics.
    var metricEntries: [MetricEntry] {
        guard let metrics else { return [] }
        return metrics.enumerated().map { index, dict in
            MetricEntry(index: index, values: dict)
        }
    }
}

// MARK: - ChartType

enum ChartType: String, Codable, Sendable {
    case bar
    case line
    case pie
}

// MARK: - MetricEntry

/// A parsed metric row extracted from StructuredData's `metrics` array.
/// Provides convenient typed accessors for common fields while preserving
/// access to the underlying raw values.
struct MetricEntry: Identifiable, Sendable, Equatable {
    let id: String
    let values: [String: JSONValue]

    init(index: Int, values: [String: JSONValue]) {
        // Use platform or label as the identifier if available; otherwise use index.
        if let platform = values["platform"]?.stringValue {
            self.id = platform
        } else if let label = values["label"]?.stringValue {
            self.id = label
        } else {
            self.id = "metric_\(index)"
        }
        self.values = values
    }

    var platform: String? { values["platform"]?.stringValue }
    var label: String? { values["label"]?.stringValue }
    var likes: Int? { values["likes"]?.intValue }
    var comments: Int? { values["comments"]?.intValue }
    var shares: Int? { values["shares"]?.intValue }
    var impressions: Int? { values["impressions"]?.intValue }
    var clicks: Int? { values["clicks"]?.intValue }
    var conversions: Int? { values["conversions"]?.intValue }
    var value: Double? { values["value"]?.doubleValue }

    /// Returns the total engagement count across all numeric metric fields.
    var totalEngagement: Int {
        (likes ?? 0) + (comments ?? 0) + (shares ?? 0)
    }

    /// Returns all non-nil key-value pairs suitable for display.
    var displayPairs: [(key: String, value: String)] {
        var pairs: [(String, String)] = []
        for (key, jsonValue) in values.sorted(by: { $0.key < $1.key }) {
            switch jsonValue {
            case .string(let s): pairs.append((key.capitalized, s))
            case .int(let i): pairs.append((key.capitalized, formatNumber(i)))
            case .double(let d): pairs.append((key.capitalized, formatDecimal(d)))
            case .bool(let b): pairs.append((key.capitalized, b ? "Yes" : "No"))
            default: break
            }
        }
        return pairs
    }

    private func formatNumber(_ value: Int) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.groupingSeparator = ","
        return formatter.string(from: NSNumber(value: value)) ?? "\(value)"
    }

    private func formatDecimal(_ value: Double) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.maximumFractionDigits = 2
        formatter.minimumFractionDigits = 0
        return formatter.string(from: NSNumber(value: value)) ?? "\(value)"
    }
}
